Add support for debug and program launch configurations

Cypress-OpenOCD opened the pull request:
### Description
As of now the only one launch configuration is generated for the Eclipse which cannot cover all use cases. This PR is intended to extend support for different launch configurations (program, debug and erase) for the "cdt" exporter.

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->
@theotherjimmy

Commit:
added support for debug and program launch configurations

cmonr reviewed (approved):
LGTM

cmonr commented:
LGTM, but afaik, we're not testing this specific exporter in CI.

@Cypress-OpenOCD Would you happen to have an example of something that this PR fixes/makes working? Something like a log or screenshot?

theotherjimmy commented:
@Cypress-OpenOCD @cmonr This exporter is named `eclipse_<toolchain>` on the command line.

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -29,11 +30,19 @@ def generate(self):
+        templates = ['%s.tmpl' % (self.target.lower())] + \
+            ['%s.tmpl' % (label.lower()) for label
+            in self.toolchain.target.extra_labels] + \
+            ['%s.tmpl' % 'pyocd_settings']
```
As there are no target-specific templates, could we use the following instead:

```python
templates = ["pyocd_settings.tmpl", "pyocd_settings_debug.tmpl", "pyocd_settings_program.tmpl"]
```

rbatyuk replied:
@theotherjimmy, I suggest to have 3 default templates:  templates = ["pyocd_settings_debug.tmpl", "pyocd_settings_program.tmpl", "pyocd_settings_erase.tmpl"] .   As I understood, the specific parameters should go through JSON file.

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -29,11 +30,19 @@ def generate(self):
+            try:
+                self.gen_file('cdt/%s' % templatefile, ctx, join('eclipse-extras',
+                           '{target}_{project}_{launch}.launch'.format(target=self.target,
+                                                                    project=self.project_name, launch=templatefile)))
```
Indentation seems off here. Could you move `target=` onto it's own line and reflow the other arguments?

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -29,11 +30,19 @@ def generate(self):
+                self.gen_file('cdt/%s' % templatefile, ctx, join('eclipse-extras',
+                           '{target}_{project}_{launch}.launch'.format(target=self.target,
+                                                                    project=self.project_name, launch=templatefile)))
+            except TemplateNotFound:
```
Just thought of this: without target-specific templates, there is no chance of missing templates. Could you get rid of the try-except here too?

rbatyuk commented on tools/export/cdt/pyocd_settings_debug.tmpl:
```diff
@@ -0,0 +1,70 @@
+<stringAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerExecutable" value="${pyocd_path}/${pyocd_executable}"/>
+<booleanAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerFlashFastVerify" value="false"/>
+<intAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerFlashMode" value="2"/>
+<intAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerGdbPortNumber" value="3334"/>
```
the default template of pyocd_settings_debug.tml should be the same as pyocd_settings.tmpl. Specific port 3334 should be in the json file

theotherjimmy replied:
@rbatyuk Could you specify which json file you're referring to?

rbatyuk commented on tools/export/cdt/pyocd_settings_debug.tmpl:
```diff
@@ -0,0 +1,70 @@
+<booleanAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerHaltAtHardFault" value="false"/>
+<stringAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerLog" value=""/>
+<!-- Remove target name(-t cy8c6xxa) once we have DAPLink support in KP3 -->
+<stringAttribute key="ilg.gnuarmeclipse.debug.gdbjtag.pyocd.gdbServerOther" value="-t cy8c6xxa&#13;&#10;-p 3333&#13;&#10;--no-deprecation-warning"/>
```
the name of target should not be included in the default template

Cypress-OpenOCD commented:
@rbatyuk, agree, there are some psoc6-related implementation in the templates. I cannot see any way to make 'generic' template without having some sort of set of target specific parameters somewhere, since some targets may require specific reset sequence etc. I am thinking about the similar approach that is used for the iar exporter. There are a set of defaults which can be redefined for the specific target (tools/export/iar/iar_definitions.json). If we make similar JSON for Eclipse we have single template per each use case (e.g. debug, program, erase) where some parameters may be added or redefined with their own flow.
@cmonr @theotherjimmy is this worth doing this?

theotherjimmy commented:
@Cypress-OpenOCD I prefer having a system like `iar_defines.json` to having multiple templates.

Cypress-OpenOCD commented:
> @Cypress-OpenOCD I prefer having a system like `iar_defines.json` to having multiple templates.

I am already working on that, looks like we could end up with a single template.

theotherjimmy commented:
> looks like we could end up with a single template.

I'm excited to see that! Will it be one template total? or 3, one each for debug/program/erase?

Cypress-OpenOCD commented:
> .
> 
> > looks like we could end up with a single template.
> 
> I'm excited to see that! Will it be one template total? or 3, one each for debug/program/erase?

Single template for all use cases. The concrete details would be implemented in the json. Hope it will work ;)

Commit:
Reworked launch configuration creation mechanism - switched to single template. Specific data comes from the JSON file

Cypress-OpenOCD commented:
Hi, please check out the reworked version of program/debug launch configurations support. In short: now there is single template for all launch configurations. Default and target-specific data comes from cdt_definitions.json. Thus, if some target require any custom sequence it may define it in the JSON file (like for CY8CPROTO_062_4343W).

cmonr commented:
Please rebase your PR.

A fix for the `travis-ci/tools-py2.7` was brought in today (https://github.com/ARMmbed/mbed-os/pull/9371) due to an update to the `hypothesis` module used in CI: https://github.com/HypothesisWorks/hypothesis/releases/tag/hypothesis-python-4.0.0

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -1,15 +1,49 @@
     """Generic Eclipse project. Intended to be subclassed by classes that
     specify a type of Makefile.
     """
+    def get_target_config(self, ctx, configuration):
```
This method does not use self. Please: 
```suggestion
    @staticmethod
    def get_target_config(ctx, configuration):
```

Cypress-OpenOCD replied:
@theotherjimmy,
It is used right here:
tgt = TARGET_MAP[self.target]

Can I resolve this one?

theotherjimmy replied:
I must have missed that. Thanks for pointing it out.

theotherjimmy reviewed (changes requested):
Minor thing, then mention me so that I can approve and we can get this improvement in.

0xc0170 commented:
> [X] Functionality change

I labeled this for 5.11.3 , please review @theotherjimmy

theotherjimmy commented:
@0xc0170 I'm already reviewing this PR.

0xc0170 commented:
> @0xc0170 I'm already reviewing this PR.

I was asking for review the label (if this can go into 5.11.3 or not)

theotherjimmy commented:
@0xc0170 Thanks for the clarification.

Yes, 5.11.3 is a fine target.

theotherjimmy reviewed (approved):
~Ya know what? I can make that change later.~ As pointed out, I missed the use of self. approval still stands.

Cypress-OpenOCD commented:
Ok, thanks.

cmonr commented:
@0xc0170 
>I was asking for review the label (if this can go into 5.11.3 or not)

The files are all focused around a specific vendor's exporter. It's not as clean as the "is the change udner TARGETS/*" soft rule that we use, but this should definitely be fine for 5.11.3.

cmonr commented:
CI started

0xc0170 commented:
One breakage on master needs to be addressed asap, the current build was aborted and will be restarted within today.

Cypress-OpenOCD commented:
Hi all, recently we have switched to the new GNU MCU Eclipse plugin. Unfortunately, Eclipse refuses to import launch configurations created with mbed export. It turns out that the launch configurations can be imported only with using former GNU ARM Eclipse plugin. Is there any plans to support new GNU MCU plugin? If you don't mind, we can add it from our side.

theotherjimmy commented:
Hi @Cypress-OpenOCD, I would like to have the launch script upgrade. I'm working on many other things, so I will take you up on your offer to provide the launch scripts.

0xc0170 commented:
CI started

cmonr commented:
Whoops. Started CI too soon, it's being stopped in the backgound.

Cypress-OpenOCD commented:
We are going to add GNU MCU Eclipse plugin support. It would be better to track the changes here.

Commit:
Added GNU MCU Eclipse plug-in support

Cypress-OpenOCD commented:
Please review the changes made to support new GNU MCU Eclipse plug-in.
Changes:
  1. Added new template to support GNU MCU Eclipse plug-in
  2. Changed naming convention for new configurations:
      - GNU ARM Eclipse - `{target}_{project}_{configuration}_pyocd_settings.launch` (remains the same)
      - GNU MCU Eclipse - `{target}_{project}_{configuration}.launch` (to distinguish from GNU ARM Eclipse)

With these changes mbed export generates two sets of launch configurations: compatible with GNU ARM Eclipse, and GNU MCU Eclipse, therefore users may use plug-in they prefer.

We are going to add few more kits on Monday, which may affect `cdt_definitions.json`.

Commit:
Fix python3 compatibility issue

Commit:
Changes:
- added new erase launch configuration
- added new kits support
- code cleanup

Cypress-OpenOCD commented:
Overview of this pull request:

1. Extended support of PyOCD launch configurations for Eclipse:
- added support of GNU MCU Eclipse plug-in (launch configurations generates now for both plug-ins)
- added 2 new launch configurations: **program** and **erase**

2. Made launch configurations template flexible and extendable:
- created `cdt_definitions.json` where launch configuration options may be extended for specific targets
- `__init__.py` uses `cdt_definitions.json` to fill the launch templates

@theotherjimmy please review

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -25,15 +56,31 @@ def generate(self):
+        launch_cfgs = {}
+        for launch_name in supported_launches:
+            launch = dict(ctx.items())
+            launch.update({'device': self.get_target_config(ctx, launch_name)}.items())
```
```suggestion
            launch.update({'device': self.get_target_config(ctx, launch_name)})
```

Commit:
Added copyright notes

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -1,18 +1,49 @@
+            
+            eclipse_config.update(target_info)
+
+        Eclipsedevice = namedtuple('Eclipsedevice', eclipse_config.keys())
```
I'm non a fan of using named tuple for this. Every return value of `get_target_config` will have a different type and this is only used inside a jinja template, which treats `foo.bar` the same as `foo[bar]`.

Cypress-OpenOCD replied:
Uses eclipse_config dictionary as a returns value

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -25,15 +56,31 @@ def generate(self):
+        for launch_name in supported_launches:
+            launch = dict(ctx.items())
+            launch.update({'device': self.get_target_config(ctx, launch_name)}.items())
+            launch_cfgs.update({launch_name: launch})
```
```suggestion
            launch_cfgs[launch_name] = launch
```

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -25,15 +56,31 @@ def generate(self):
+        for launch_name, ctx in launch_cfgs.items():
+            # Generate launch configurations for former GNU ARM Eclipse plug-in
+            self.gen_file('cdt/%s' % 'pyocd_settings_gnu_arm.tmpl', ctx, join('eclipse-extras',
+                            '{target}_{project}_{conf}_{launch}.launch'.format(
```
```suggestion
                            '{target}_{project}_{conf}_pyocd_settings.launch'.format(
```

And drop the `launch='pyocd_settings'` below

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -1,18 +1,49 @@
+        """Retrieve info from cdt_definitions.json"""
+        tgt = TARGET_MAP[self.target]
+        defaults = _CONFIGS_OPTIONS['default']
+        eclipse_config = defaults['generic']
```
Since you modify this, could you please copy it?

Cypress-OpenOCD replied:
I copied all of these dictionaries, correct?

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -1,18 +1,49 @@
+
+        target_specific = _CONFIGS_OPTIONS['targets']
+        if tgt.name in target_specific:
+            target_info = target_specific[tgt.name]['generic']
```
This is modified, so multiple exports on the online compiler will lead to unintended behavior when they all modify this variable over and over again.
```suggestion
           eclipse_config.update(target_specific[tgt.name]['generic'])
```

If you change ^ and you replace `target_info` with `eclipse_config` 2 lines down, you achieve the same affect without having to modify more dictionaries.

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -25,15 +56,31 @@ def generate(self):
+        
+        launch_cfgs = {}
+        for launch_name in supported_launches:
+            launch = dict(ctx.items())
```
This is an odd way to copy. Could you please use the more direct:
```suggestion
            launch = deepcopy(ctx)
```

You will  have to `from copy import deepcopy` at the top of the file.

theotherjimmy reviewed (changes requested):
Most of these comments are style nits or online compiler tweaks. The online compiler will load this file _once_ and use it for many, many exporters. Therefore it is important to make sure that you don't modify any global state within the exporter to avoid accidentally copying settings between users exporting.

theotherjimmy committed:
Apply suggestions from code review

Co-Authored-By: Cypress-OpenOCD <39907069+Cypress-OpenOCD@users.noreply.github.com>

Commit:
Resolved code review comments

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -1,18 +1,63 @@
     """
+    def get_target_config(self, ctx, configuration):
+        """Retrieve info from cdt_definitions.json"""
+        tgt = deepcopy(TARGET_MAP[self.target])
```
I don't think you need this deepcopy. Further, I think you only use `tgt.name` which is always the same as `self.target`.

theotherjimmy commented on tools/export/cdt/__init__.py:
```diff
@@ -1,18 +1,63 @@
+    def get_target_config(self, ctx, configuration):
+        """Retrieve info from cdt_definitions.json"""
+        tgt = deepcopy(TARGET_MAP[self.target])
+        defaults = deepcopy(_CONFIGS_OPTIONS['default'])
```
I don't think you modify defaults, so this copy may not be needed.

Cypress-OpenOCD replied:
Ok, I think this may be kept to prevent any issues in case future updates

theotherjimmy reviewed (approved):
Looks good. Previous comments are not blocking.

cmonr commented:
CI started
